utils/reporter_project_query.py

```diff
@@ -0,0 +1,158 @@
+import json
+import argparse
+import requests
+import numpy as np
+import math
+import pandas as pd
+
+
+def get_args():
+
+    parser = argparse.ArgumentParser(
+        description="Access and validate tables from Synapse"
+    )
+    parser.add_argument(
+        "-g",
+        nargs="+",
+        help="Space separated list of grant numbers to query NIH RePORTER",
+    )
+    parser.add_argument(
+        "-y",
+        nargs="+",
+        default=[],
+        help="The list of years to use when searching for grant information.",
+    )
+    parser.add_argument("-c", help="Path for CSV where converted JSON will be stored.")
+
+    return parser.parse_args()
+
+
+def build_payload(grant_numbers, years, lim):
+
+    criteria_dict = dict(
+        project_nums=grant_numbers, sub_project_only=True, fiscal_years=years
+    )
+
+    include = [
+        "ProjectTitle",
+        "ProjectNum",
+        "SubprojectId",
+        "FiscalYear",
+        "ProjectEndDate",
+    ]
+
+    payload = dict(
+        criteria=criteria_dict,
+        include_fields=include,
+        offset=0,
+        limit=lim,
+        sort_field="fiscal_year",
+        sort_order="desc",
+    )
+
+    json_payload = json.dumps(payload)
+
+    return json_payload
+
+
+def get_reporter_info(search_payload, header_content):
+
+    req = requests.post(
+        url="https://api.reporter.nih.gov/v2/projects/search",
+        headers=header_content,
+        data=search_payload,
+    )
+
+    return req
+
+
+def build_report(report_list, out_path):
+
+    df_list = []
+    for report in report_list:
+        report_df = pd.read_json(report, orient="records")
+        df_list.append(report_df)
+
+    full_report = pd.concat(df_list).reset_index(drop=True)
+
+    print(f"\n\nYour report is available at {out_path}.\n\n")
+
+    full_report.to_csv(out_path, index=False)
+
+
+def main():
+
+    args = get_args()
+
+    in_grants, in_years, csvPath = args.g, args.y, args.c
+
+    grants = []
+
+    years = []
+
+    for grant in in_grants:
+        grant = "".join([grant, "*"])
+        grants.append(grant)
+
+    for year in in_years:
+        year = int(year)
+        years.append(year)
+
+    headers = {"content-type": "application/json"}
+
+    max_request_length = int(400)
+    grant_count = int(len(grants))
+    req_per_grant = int(20)
+
+    if grant_count > max_request_length / req_per_grant:
+        split_count = math.ceil(grant_count / req_per_grant) + 1
+        print(
+            f"Due to API restrictions, your query will be split into {split_count} parts"
+        )
+        grant_chunks = [g.tolist() for g in np.array_split(grants, split_count)]
+
+    else:
+        grant_chunks = grants
```

Comment:
I noticed that `grant_chunks` will either be a list of lists or it'll be a list of strings, depending on the number of grants to be queried --  I wanted to make sure this was intentional?

For example, for the `if` code chunk, I used the list of grant numbers from [grantview_synapse_storage_manifest_table](https://www.synapse.org/Synapse:syn53259587/tables/) and got a list of lists in return:

```python
>>> grants
['CA184897*', 'CA184898*', ..., 'CA284086*', 'CA284090*']
>>> split_count = math.ceil(grant_count / req_per_grant) + 1
>>> grant_chunks = [g.tolist() for g in np.array_split(grants, split_count)]
>>> grant_chunks
[['CA184897*', 'CA184898*', ...], ..., ['CA274504*', 'CA274506*',...]]
```

For the `else`, I used a smaller subset and got back a list of strings:

```python
>>> grant_chunks2 = grants_subset
>>> grant_chunks2
['CA184897*', 'CA184898*']
```

Reply:
Thanks for pointing this out. I definitely wanted a list of lists in both cases, so I changed the else statement to `grant_chunks = [grants]`